### ISSUE TITLE
Merging of a new plotting module into SymPy

### ISSUE THREAD

**Issue by Davide-sd** [sympy/sympy#23036]
A few months ago NumFocus selected my proposal to implement a new plotting module for SymPy. Now that the development is done, @oscarbenjamin  and @asmeurer suggested to open this issue in order to understand if/how this module can be integrated into SymPy.

## Proposal - Objectives

The aim of this proposal is to integrate different new open-source technologies. In particular:

* Integration of different plotting libraries: Matplotlib, Bokeh, Plotly, K3D-Jupyter, etc. where the latters are native-interactive. This would allow the user to choose the best library for the specific problem (or based on the user preference).
* Integration of holoviz’s [panel](https://github.com/holoviz/panel/) and [param](https://github.com/holoviz/param) libraries: together with the aforementioned native-interactive plotting libraries, this enables the creation of a function, `iplot` (interactive-plot), to visually explore symbolic expressions by changing the parameters using widgets (sliders, checkboxes, buttons, etc.) with a single command and virtually no knowledge of panel/param.
* Add a function to easily and quickly extract numerical data from the supported types of plot in order to create custom plots. The current procedure is manually-intensive and must be adapted to the specific case.
* Add functions to quickly visualize 2D/3D vector fields, complex numbers and complex functions. Integrate them with the `iplot` function.
* Refactoring of the `MatplotlibBackend`: the current implementation (inherited from SymPy) is blocking its integration with the `iplot` function.
* Writing all the necessary documentation and examples in order to take full advantage of the module.
* Improve singularity handling for plots (SymPy issue #18455).
* Improve support for plotting Piecewise functions (the discontinuities must be visible, at least in 2D plots).

## Results

The new plotting module is available at [this Github repository](https://github.com/Davide-sd/sympy-plot-backends). All the previous plotting functions are available, as well as new ones:

* `plot_list`: add custom numerical data to the plot. 
* `plot_polar`: create polar plots...
* `plot_geometry`: plot entities from `sympy.geometry`
* `plot_piecewise`: visualize `Piecewise` expressions with their discontinuities.
* `plot_vector`: visualize 2D/3D vector fields with quivers or streamlines.
* `plot_complex`, `plot_complex_list`, `plot_real_imag`, `plot_complex_vector` deal with complex functions.
* `iplot` is a general function that allows to create interactive plots with widgets (sliders, checkboxes, ...) starting from symbolic expression.
* `plotgrid` to combine multiple plots into a grid-like layout.

By looking at [the documentation](https://sympy-plot-backends.readthedocs.io/en/latest/) you'll be able to explore the new (and old) plotting functions as well as their output.

A few notebook tutorials are available in the repository: they are designed to illustrate the capabilities and the differences between the backends (the plotting libraries). Sadly, they must be executed locally, as Nbviewer is unable to load all javascript-based plots.

## Back Compatibility (or lack of)

While the plotting functionalities are overall similar to the current `sympy.plotting`, there are some important differences:

* In `sympy.plotting`, `Plot` and `MatplotlibBackend` are unrelated (in OOP terms). The instances of these two classes create a cross reference. In this new module, `Plot` serves as a base class to `MatplotlibBackend` (and every other backend).
* In `sympy.plotting`, each `*Series` class has its own method to return numerical data (for exampe, `get_data()`, `get_meshes()`, `get_raster()`). In the new module all series only exposes `get_data()`.
* `sympy.plotting` exposes the `PlotGrid` class. This has been removed in favor of the `plotgrid` function, which is also able to combine plots created with different backends.
* The adaptive algorithm is also different: this module relies on [adaptive](https://github.com/python-adaptive/adaptive/), which allows more flexibility.
  * The `depth` keyword argument has been removed, while `adaptive_goal` and `loss_fn` have been introduced to control the new module.
  * It has also been implemented to 3D lines and surfaces.
  * It allows to generate smoother line plots, at the cost of performance.
* `sympy.plotting` exposed the `nb_of_points_*` keyword arguments. These have been replaced with `n` or `n1, n2`.
* `sympy.plotting` exposed the `label`, `line_color` and `surface_color` keyword argument (both of plotting functions and data series). These have been removed in this module. It means that with `sympy.plotting` we could modify the appearance of a plot after a `Plot` object was created. This is not supported with the new module: all customization options must be provided when a plotting function is called.
* The `plot_implicit` function has different signature. Moreover, by default it uses a mesh grid algorithm and contour plots (in contrast to the adaptive algorithm). It is going to automatically switch to an adaptive algorithm if Boolean expressions are found. This ensures a better visualization for non-Boolean implicit expressions.
* Almost completely removed the dependency on `sympy.plotting.experimental_lambdify` in favor of `lambdify`. Complete removal will be achieved if #23010 will be solved.

## Testing

Several tests have been implemented on this new module. In particular:

* All testing available in `sympy.plotting` has been incorporated (some tests have been modified to account for the differences of this new module).
* Added tests to assure that `*Series` classes correctly generates data.
* Added tests to assure that plotting functions are able to pre-process their arguments (for example, when plotting piecewise expressions or summations).
* Added tests to assure that backends provide a common user experience. 

## Discussion

There are many things to discuss:

* Notably, if this new modules is going to be merged into SymPy, how do we deal with the old plotting module? Remember, all plotting functions available on `sympy.plotting` are also available in the new module.
* If you have specific questions about why I changed something, I'll provide a detailed answer.
* Install the new module and try it. Read the documentation, follow the tutorials. If you find bugs or if you think the documentation can be improved, please let me know either with an issue or a comment down below.

**Comment by hanspi42** [comment]
I looked at one tutorial (complex plots) and like the new module a lot! I do not know very much about plotting from SymPy (I always use lamdify/matplotlib; maybe this will change in the future with this work), so maybe the following is trivial: it would be wonderful to be able to change the horizontal and/or vertical axis to logarithmic.

**Comment by moorepants** [comment]
I looked briefly at your docs and it looks very nice. My only comment for merging into sympy is that it would be best if users can use the existing `plot()` function to get your results. This would be the third sympy plotting module and having new API for each one is confusing and complicated for users. Calling `sympy.plot()` on expressions should just work, IMHO. If there is some way to make a transition such that the user doesn't have to change their behavior or existing code, that's ideal.

**Comment by Davide-sd** [comment]
> I looked at one tutorial (complex plots) and like the new module a lot! I do not know very much about plotting from SymPy (I always use lamdify/matplotlib; maybe this will change in the future with this work), so maybe the following is trivial: it would be wonderful to be able to change the horizontal and/or vertical axis to logarithmic.

You can use `xscale="log"` and `yscale="log"` to control the two axis (by default, they are set to `"linear"`). Thanks for pointing this out, as I forgot to document this keyword arguments on the complex module!

**Comment by Davide-sd** [comment]
> Does this imply that SymPy would need a dependency on adaptive?

It would be a soft-dependency on the user side, same as it is now with matplotlib and numpy. In order to use the plotting module, the user has to installed them manually.

> If there is some way to make a transition such that the user doesn't have to change their behavior or existing code, that's ideal.

> I'm not entirely sure how these are used but it should be easy enough to provide a backwards compatible option here and just document it as being a deprecated way to call the functions.

This could be achieved for some keyword arguments, like `nb_number_of_points_*`, but not for others, like `line_color` or `surface_color`. This is because the two modules are structurally different. For example, with sympy.plotting  `line_color` and `surface_color` are attributes of `*Series` classes and `MatplotlibBackend` was specifically coded to look for their values. The new module doesn't implement such specific attributes: this slender approach allows for the full control of the appearance of the plot with any backend, so I could set the line color, line opacity, line style, ..., without having to create all these attributes.

In this regard, full back-compatibility is impossible!

**Comment by oscarbenjamin** [comment]
> The new module doesn't implement such specific attributes

Can you expand on this? What specifically is the code that a user would use now to do these things? What would they need to change it to? What happen if they didn't change their code (do they see an exception or is their code silently ignored)?

**Comment by Davide-sd** [comment]
> Can you expand on this? What specifically is the code that a user would use now to do these things? What would they need to change it to? What happen if they didn't change their code (do they see an exception or is their code silently ignored)?

With sympy.plotting there are two ways to change line color. Let's compare sympy.plotting with my module:

### First way - at function call

With sympy.plotting, the following shows a plot in which the line color is red.
```python
from sympy import *
var("x, y")
plot(sin(x), line_color="red")
```
With my module, the previous command shows a plot in which the line color is blue. That's because `line_color` is not recognized. Instead, the user has to provide the `line_kw` dictionary, containing backend-specific keyword arguments to customize the plot:

```python
from sympy import *
from spb import plot
var("x, y")
plot(sin(x), line_kw=dict(color="red", linestyle="--", alpha=0.5))
```

Now, the plot shows a dashed red line with opacity 0.5. In the previous command, `color="red", linestyle="--", alpha=0.5` are passed directly to matplotlib's `ax.plot` function. If we were using Plotly, we would use something different.

### Second way - setting the series attribute after the function call

With sympy.plotting it is possible to modify the series attributes after the plot object has been created.

```python
from sympy import *
var("x, y")
p = plot(sin(x), cos(x), show=False)
p[1].line_color = "red"
p.show()
```
This will show a blue line and a red line. 

Running the previous code with my module, the figure would show a blue line and an orange line (again, no errors). That's because the series don't have the `line_color` attribute, and the backends are not looking for it. If I want to set a specific color only on a specific data series, the procedure is the following:

```python
from sympy import *
var("x, y")
from spb import plot
p1 = plot(sin(x), show=False)
p2 = plot(cos(x), line_kw=dict(color="red"), show=False)
(p1+p2).show()
```

Within sympy.plotting the `surface_color` attribute/keyword argument allows to apply a coloring function to a surface. For example:

```python
from sympy import *
var("x, y")
plot3d(sin(x)*y, (x, -3*pi, 3*pi), (y, -5, 5), surface_color=lambda a, b: np.sqrt(a**2+b**2))
```
Here, the colors is going to change radially from the center, according to a fixed color map, which is impossible to change. With this approach, there is no possibility to set a custom solid color to the surface.

With my module, setting a color function is harder, in favor of the capability to easily change a colormap or set a solid color. For example:

```python
import matplotlib.cm as cm
# change the colormap
p1 = plot3d(cos(x**2 + y**2), (x, -2, 2), (y, -2, 2), surface_kw=dict(cmap=cm.Reds))

# apply a custom solid color
p2 = plot3d(cos(x**2 + y**2), (x, -2, 2), (y, -2, 2), use_cm=False, surface_kw=dict(color="green"))

# apply a color function
import numpy as np
p3 = plot3d(sin(x)*y, (x, -3*pi, 3*pi), (y, -5, 5), show=False)
xx, yy, zz = p3[0].get_data()
colors = np.sqrt(xx**2 + yy**2)
colors = cm.viridis(colors / colors.max())
plot3d(sin(x)*y, (x, -3*pi, 3*pi), (y, -5, 5), surface_kw=dict(facecolors=colors))
```

Here, the content of `surface_kw` is sent directly to matplotlib's `ax.plot_surface` function. If we were to use a different backend, the content of this dictionary would be different.

In the new module, these rendering related dictionaries, such as `line_kw` and `surface_kw` allowed for the creation (and customization) of "composite" plotting functions, such as `plot_vector`, in which there could be contours, quivers and streamlines all in the same plot. So, customizing them is really easy, no matter the backend.

**Comment by oscarbenjamin** [comment]
> With sympy.plotting it is possible to modify the series attributes after the plot object has been created.

Is this a fundamental limitation of the new design? I agree that assigning the `p[1].line_color` isn't a good API but it is documented and I'm sure that someone is using it. I'd guess that a setter method could be added to `p[1]` so that it can update the colour at least for the matplotlib backend. Perhaps the setter method could give a warning suggesting to pass the colour in a different way that works with all versions of sympy.

> With my module, the previous command shows a plot in which the line color is blue. That's because `line_color` is not recognized.

Presumably this could be changed to accept a `line_color` kwarg and add it to `line_kw` so that this option continues to work.

> Instead, the user has to provide the `line_kw` dictionary , containing backend-specific keyword arguments to customize the plot:

Why can't the things from `line_kw` just be passed directly as kwargs from `plot` to the backend. Why does there need to be a separate `line_kw` parameter to take these?

I don't see why the API is this:
```
p2 = plot(cos(x), line_kw=dict(color="red"), show=False)
```
Rather than this:
```
p2 = plot(cos(x), color="red", show=False)
```

**Comment by asmeurer** [comment]
> This is not supported with the new module: all customization options must be provided when a plotting function is called.

This actually seems concerning to me. I don't know about other plotting libraries, but with matplotlib, it is a very standard thing to modify the plot attributes after the plot is created. Quite a few things would be impossible to do without this, e.g., integrating a plot of a SymPy function into a larger plot.

Ideally, though, users should do this by just accessing the underlying plot object and modifying it directly using matplotlib functions. This means they have to use backend specific functions, but it also means that we don't have to include abstractions for every specific detail of a plot in our plotting API.

So, at least for matplotlib, what I would like to see is something like

```
p = plot(sin(x))
```

**Comment by Davide-sd** [comment]
> Why can't the things from `line_kw` just be passed directly as kwargs from plot to the backend. Why does there need to be a separate `line_kw` parameter to take these?

The main reason is code simplicity.

The backend doesn't know from which plotting function it has been instantiated. All a backend knows is that there are some data series that have some attributes that must be plotted. The backend contains just a loop that iterates over data series. Essentially, the backend just reduces code repetition that would be otherwise scattered into any plotting function.
 
Some plotting functions, like `plot_vector`, combines multiple data series into the same figure, for example contours, quivers, lines, ... Example: suppose I have a function creating a contour  + a line. Let's say I didn't use `line_kw` (or other similar dictionaries), and that I would like to change the colormap used by contours. Then, I would execute `plot_function(expression, cmap=cm.Reds)`. At this point, `cmap` would be sent to `ax.contours` as well as `ax.plot`, which would raise an `AttributeError`. To solve this issue, I would need to check every single keyword argument that gets into any backend's specific plot function, for example I would need to check that `cmap` doesn't get into `ax.plot`. That is a nightmare to implement, and it would also be a stupid thing to do, as plotting libraries might evolves over time, deprecating some keyword argument while allowing something new.

Moreso, some plotting library might use keyword arguments that may already be used by SymPy plotting interface: I don't have any example of this yet, but since I'm supporting multiple plotting libraries, the changes of that happening are not so small... 

Dictionaries like `line_kw`, `surface_kw`, ..., solves all these problems.

> This actually seems concerning to me. I don't know about other plotting libraries, but with matplotlib, it is a very standard thing to modify the plot attributes after the plot is created.

Do not confuse things:

* data is being generated by data series.
* the backend is a convenient class that plots data.
* the figure object is an attribute of the backend, it can be easily accessed with the `fig` attribute.

So, let me give you an example: you would like to change the appearance of a simple line plot (change line style, color, opacity).

Method 1: using the plotting function

```python
from sympy import *
from spb import plot
var("x")
plot(sin(x), line_kw=dict(color="magenta", linestyle="--", alpha=0.5))
```

Method 2: retrieving the figure object

```python
from sympy import *
from spb import plot
var("x")
p = plot(sin(x))
p.fig.axes[0].lines[0].set_color("magenta")
p.fig.axes[0].lines[0].set_linestyle("--")
p.fig.axes[0].lines[0].set_alpha(0.5)
```

This is easy, right? Well, I challenge you: try to change the colormap of a figure created with `plot_contour` in matplotlib using method 2. All of a sudden, my implementation with `line_kw`, `surface_kw`, ..., makes sense, because some matplotlib objects are extremely hard or nearly impossible to modify after a figure has been created.

Moreso, the approach with `line_kw`, ..., allows the user to explore the documentation associated to each plotting library, and use the keyword arguments listed on those pages. Method 2, instead, relies on the user to have a good knowledge of OOP. Some plotting libraries makes it easy to modify the plot after it has been created (for example, Plotly), other libraries are much more convoluted (Matplotlib and Bokeh).

**Comment by asmeurer** [comment]
Say I have an existing figure that I've made entirely with matplotlib, and I want to add a function plot generated by SymPy to it? How easy is it do that? Can you extract relevant parts of the SymPy plot and inject them into the existing matplotlib figure? Or would it be better to just get the SymPy functions to tell you data series and let you re-plot them? 

I think it something like this is doable then that's a sign that the module has a good level of extensibility.

**Comment by oscarbenjamin** [comment]
I know I've said this a few times before but personally I think that the API for sympy's plotting module should be something like
```
plot(expr, (z, 0, 5)) # simple case, only handles a few options
```
Of course we can have some very nice plotting functions like this. They just don't need to be super customisable because if someone wants to taylor the output then they could use the underlying library directly. For more complicated things it should be:
```
import matplotlib.pyplot as plt
x, y = plot_xy(expr, (z, 0, 5))
plt.plot(x, y)
# call any matplotlib routines you want to customise the plot
```
What is lacking is the `plot_xy` function that gives the data that you can just pass directly to matplotlib. As soon as a user has that then a whole world of plotting possibility is opened up that sympy's plotting module could never compete with.

Other convenience functions could be added but I'm wary of the approach that the `sympy.plot` function signature needs to be the gateway to all the advanced functionality of the backends. We should definitely make it easier for someone to just get the `x` and `y` arrays that would be passed directly to matplotlib though.

The approach with `line_kw` is strange because the options specified in`line_kw` are backend-specific but they all need to pass through the `sympy.plot` function. Why not just make it easier to use the backend directly and then sympy.plot can just handle a very limited set of options? I don't see the advantage of the "sympy.plot with multiple backend" architecture if it doesn't provide a uniform interface over the different backends.

**Comment by Davide-sd** [comment]
> Say I have an existing figure that I've made entirely with matplotlib, and I want to add a function plot generated by SymPy to it? How easy is it do that? Can you extract relevant parts of the SymPy plot and inject them into the existing matplotlib figure? Or would it be better to just get the SymPy functions to tell you data series and let you re-plot them?

As far as I know, matplotlib does have a strong relationship between the figure and its axes, so it doesn't allow to take parts of an axis and copy it to a new figure. Also, the backend can't take an existing matplotlib figure, as it requires a 1-to-1 relationship between what's in the figure and its data series. If a an existing figure/axis would be given to the backend, its original data would be wiped. The backend will show a figure everytime its `show()` method is called. If data wouldn't be wiped and you were to call `show()` multiple times, you would add again and again the same data series to the plot, making it unreadable (legend will trigger on, ...).

Which means you will have to use one of the plot functions to generate the numerical data. Here is an example:

```python
import numpy as np
import matplotlib.pyplot as plt
xx1 = np.linspace(-3, 3)
yy1 = np.cos(xx1)

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
ax.plot(xx1, yy1, label="numpy")

p = plot(sin(x) * cos(x), (x, -3, 3), show=False)
xx2, yy2 = p[0].get_data()

ax.plot(xx2, yy2, label="sympy")
ax.legend()
plt.show()
```

> What is lacking is the `plot_xy` function that gives the data that you can just pass directly to matplotlib. As soon as a user has that then a whole world of plotting possibility is opened up that sympy's plotting module could never compete with.

I didn't went with `plot_xy`, or `plot_contour_data`, or `plot_vector_data` because they would just be wrappers that would index a data series and run `p=plot(sin(x)); xx, yy=p[0].get_data()`.

Instead, I tried to implement a more general function, `get_plot_data`. I also deprecated it in favor of `p=plot(sin(x)); xx, yy=p[0].get_data()`. Here is the story.

In a perfect world, you would throw any expression to `plot` and it would be able to visualize it. Let's understand the problem with a few expressions:

* `sin(x)`: easy, one expression with one free symbol, it can only be a line plot over the range (x, -10, 10), for example.
* `cos(x**2 + y**2)`: one expression with two free symbols. It can only be a surface/contour plot. Here, the user must also provide the ranges in order to sort out the order, otherwise the function might return data for the x axis that was supposed to be in y axis.
* `sin(x), cos(x)`: two expressions with one free symbol. What are you trying to get? Two line plots? Or a parametric plot? Let's decide that this is a case for a parametric plot.
* `z, y, x`: what is it? Is it a 3D vector field, is it something else?
* `cos(x) + sin(I * x)`: one expression with one free symbol. Should be a line plot, but I would like to get its absolute value and its argument, similar to what `plot_complex` does.... So we would have to provide keyword arguments to control the behavior. Now, we are getting deep into this rabbit hole...

The above approach only works if the user provides all the expected ranges, and the correct keyword arguments. If the user fails to provide even one range, or provided a wrong keyword argument, the function doesn't work, or worse, it works but it might returns data associated to a different type of plot, without even knowing it.

On the opposite side, all plot functions knows exactly what they are supposed to visualize. For example, `plot` expects one expression with one free symbol. If you provide something different, it raises an error. Similarly, `plot_parametric` expects two expressions with one free symbol. If it receive something different it raises an error...

Ultimately, by having multiple specific plotting functions we can retrieve the data by indexing a data series and calling `p[0],get_data()`. Here, you are absolutely certain that what you get is the correct data. In fact, you can even visualize it before retrieving it.

At the and, I deprecated the `get_plot_data` function because the plot interface allows to do that in a better and more reliable way by indexing the data series.

> The approach with `line_kw` is strange because the options specified in `line_kw` are backend-specific but they all need to pass through the sympy.plot function. Why not just make it easier to use the backend directly and then sympy.plot can just handle a very limited set of options? I don't see the advantage of the "sympy.plot with multiple backend" architecture if it doesn't provide a uniform interface over the different backends.

I understand that you don't care at all about backends. Fine. As an example, let's say you would like create a contour plot starting from a symbolic expression. As you don't care about the backend, you would like to do something like:

```python
from sympy import *
import matplotlib.pyplot as plt
import matplotlib.cm as cm
var("x, y")
xx, yy, zz = get_contour_data(cos(x**2+y**2))
plt.figure()
plt.contourf(xx, yy, zz, cmap=cm.Reds, **kwargs) # kwargs to further customize the appearance
plt.show()
```

The backend job is to deal with all the boilerplate code, so you wouldn't have to repeat it every time you need a plot. This was true for sympy.plotting and it also true for my module. To me, this is an essential feature: the less time I spend writing boilerplate code, the better I live. Now, with this module, you'd do something like this:

```python
plot_contour(cos(x**2 + y**2), (x, -2, 2), (y, -2, 2), contour_kw=dict(cmap=cm.Reds, levels=np.linspace(0, 1, 5)))
```

Everything inside `contour_kw` gets passed directly to matplotlib's `ax.contourf`, so you can fully customize the output. Why do I need this? Turns out that it is impossible to change the appearance of a contour plot (with matplotlib) after the contour has been created, or if it's possible, I still haven't found the procedure, nor I care because it would be a tremendous PITA!

In my opinion, the `line_kw`, `surface_kw`, ..., approach is the ultimate abstraction of the plotting function in relation to the backend. Plotting function and data series doesn't need to know anything about the backend: this is true for this module. However, a plot function should allow to customize the output, but not in a very limited way, as it was done in sympy.plotting, otherwise you would get people asking for and adding more options. It must provide a way to harness the full potential of the plotting library being used. With this module, plot functions are unaware of `line_kw`, `surface_kw`, ..., they just get passed down to the backend.

> I don't see the advantage of the "sympy.plot with multiple backend" architecture if it doesn't provide a uniform interface over the different backends.

This is precisely what `line_kw`, `surface_kw`, ..., achieves: a uniform interface over different backends. What goes into `line_kw`, `surface_kw`, ..., depends on the backend being used! If you fill them with matplotlib-specific commands and use it with Plotly, obviously errors will be raised.  But if you use Plotly, than you should provide Plotly-specific commands!!! By doing so, you have full customization options of every output with every backend.

**Comment by oscarbenjamin** [comment]
> In fact, you can even visualize it before retrieving it.

I take it you mean that I can look at the plot before extracting the data that made the plot. Is that a correct interpretation? What if I don't want to see the plot before getting the data?

**Comment by Davide-sd** [comment]
> > In fact, you can even visualize it before retrieving it.
> 
> I take it you mean that I can look at the plot before extracting the data that made the plot. Is that a correct interpretation? What if I don't want to see the plot before getting the data?

Correct. If you don't want to show it, you just pass `show=False` on the function call.

**Comment by oscarbenjamin** [comment]
> If you don't want to show it, you just pass `show=False` on the function call.

Does that still create the matplotlib figure behind the scenes though (even if it isn't shown)?

Suppose I want to create a figure using matplotlib that has two subplots. For one of those subplots I want to use sympy's plotting module but for the other I want to use some other code. How would I do that with the proposed API?

**Comment by Davide-sd** [comment]
> Does that still create the matplotlib figure behind the scenes though (even if it isn't shown)?

No, it is only created once the `show()` method gets executed.

> Suppose I want to create a figure using matplotlib that has two subplots. For one of those subplots I want to use sympy's plotting module but for the other I want to use some other code. How would I do that with the proposed API?

It is actually possible, but it is undocumented and it will stay so. If I were to document it, then people would expect `plot()` to also accept any matplotlib figure with data already in it, but that can't happen, because the backend is going to wipe the axis clean before looping over its series (this is because of the requirement for the backend of a 1-to-1 correspondence between its data series and what's in the figure).

So the only way to do what you want is to extract the numerical data with the procedure aforementioned and add it to your axis.

**Comment by oscarbenjamin** [comment]
Okay so there are two clear options here:

1. Call `plot(sin(x), (x, 0, 10), line_kw={...})
2. Call `plot(sin(x), (x, 0, 10), show=False).get_data()` and then use that with native matplotlib APIs directly.

If I want to do something like set the tick-labels explicitly do I have to use option 2 and handle the associated boiler plate?

**Comment by Davide-sd** [comment]
> Okay so there are two clear options here:

Correct.

> If I want to do something like set the tick-labels explicitly do I have to use option 2 and handle the associated boiler plate?

Or you can do `p = plot(sin(x), (x, 0, 10)`, then retrieve the figure object with `p.fig` and from there you have access to the axis object and you can do the necessary edits.

**Comment by oscarbenjamin** [comment]
Okay I thought you were saying above that `p.fig` was not possible. So there are three options:

1. Call `plot(sin(x), (x, 0, 10), line_kw={...})`
2. Call `plot(sin(x), (x, 0, 10), show=False).get_data()` and then use that with native matplotlib APIs directly.
3. Call `p = plot(sin(x), (x, 0, 10), line_kw={...})` and use `p.fig` for further customisation.

Under option 3 presumably the `p.fig` doesn't exist if `show=False`. How would I prevent the figure from showing to screen before calling `ax.set_xticks()`? Also how would I access `ax` from `p.fig`?

I haven't used SymPy's plotting module much myself so maybe this is obvious but how do I save a plot to a file without showing it on screen?

**Comment by Davide-sd** [comment]
> Under option 3 presumably the `p.fig` doesn't exist if `show=False`. How would I prevent the figure from showing to screen before calling `ax.set_xticks()`? Also how would I access `ax` from `p.fig`?

When you execute `p = plot(sin(x), (x, 0, 10), show=False)`, no figure is created and thus no figure is shown on the screen. However, if you then call `p.fig` the backend is going to build the figure, add numerical data and returns the figure. If you are running in Ipython or a python shell, no figure is shown on the screen. If you are running on Jupyter Notebook, the figure will be shown on the screen: this is a behavior of Matplotlib/Jupyter, can't be changed...

> I haven't used SymPy's plotting module much myself so maybe this is obvious but how do I save a plot to a file without showing it on screen?

Again, two options:

1. use `p.save(filename, **kwargs)` which is just a wrapper to matplotlib `savefig` method.
2. returns matplotlib's figure and call `savefig`.

**Comment by oscarbenjamin** [comment]
Where are we at with this?

I'd like to get this in for the next release if possible.

**Comment by Davide-sd** [comment]
In the last couple of months I had the chance to really use this module, so I have identified some features that I'm going to implement. Some are relatively easy, other might take some time.

Anyway, the most pressing issues that block the merging are #23064 and #23010

When is the next release scheduled?

**Comment by asmeurer** [comment]
New features shouldn't be a blocker to merge this unless you anticipate them changing the API or some other kind of compatibility break.

**Comment by Davide-sd** [comment]
@asmeurer @oscarbenjamin 

Since the last message I have:

* Removed the dependency on `sympy.plotting.experimental_lambdify` in favor of `lambdify`.
* Added support for polar plots, 3D implicit plots and polar discretization for 3D plots.
* Re-added the `label`, `line_color` and `surface_color` keyword argument.

Let's recap the differences between the current `sympy.plotting` and the new module:

* In `sympy.plotting`, `Plot` and `MatplotlibBackend` are unrelated (in OOP terms). The instances of these two classes create a cross reference. In this new module, `Plot` serves as a base class to `MatplotlibBackend` (and every other backend).
* In sympy.plotting, each `*Series` class has its own method to return numerical data (for exampe, `get_data()`, `get_meshes()`, `get_raster()`). In the new module all series only exposes `get_data()`.
* `sympy.plotting` exposes the `PlotGrid` class. This has been removed in favor of the `plotgrid` function, which is also able to combine plots created with different backends.
* The adaptive algorithm is also different: this module relies on [adaptive](https://github.com/python-adaptive/adaptive/), which allows more flexibility.
  * The `depth` keyword argument has been removed, while `adaptive_goal` and `loss_fn` have been introduced to control the new module.
  * It has also been implemented to 3D lines and surfaces.
  * It allows to generate smoother line plots, at the cost of performance.
* `sympy.plotting` exposed the `nb_of_points_*` keyword arguments. These have been replaced with `n` or `n1, n2`.
* The `plot_implicit` of the new module uses a mesh grid algorithm and contour plots by default (in contrast to the adaptive algorithm of `sympy.plotting`). It is going to automatically switch to an adaptive algorithm if `Boolean` expressions are found. This ensures a better visualization for non-`Boolean` implicit expressions.

I'm ready to attempt to merge the new module into SymPy. Do you have any recommendation on how to proceed?

**Comment by asmeurer** [comment]
I guess a question is how we want to handle backwards compatibility. Do we want to keep the old APIs around, but deprecated? And would doing so require keeping the entire old module around? I don't think we need to worry about change that result in different visual output, but API changes like changed keyword argument names are a bigger deal because they will break existing code.

**Comment by oscarbenjamin** [comment]
I'm presuming that this work involves changing the plotting module rather than completely replacing it with a new module. If that's the case then my suggestion is:

1. Do the necessary backwards incompatible changes first so we can clearly understand what they are before anything else happens.
2. Then do the refactorings like changing which class subclasses which other class but make these PRs pure refactoring with no intended behavioural changes.
3. Then add new features. This part could in principle be a single large PR. I would recommend trying to break it down a bit though to make it easier to review though.

**Comment by Davide-sd** [comment]
> Do the necessary backwards incompatible changes first so we can clearly understand what they are before anything else happens.

I've looked at the Deprecation Policy, but I have a question. How do I proceed with attributes and methods that have been removed?

**Comment by oscarbenjamin** [comment]
Can you give an example of an attribute being removed? Is it definitely a public attribute?

It is not always necessary to do the deprecation process and in some cases it's just not practically possible.

**Comment by Davide-sd** [comment]
> Can you give an example of an attribute being removed? Is it definitely a public attribute?

The following attributes of the current `Plot` class will be removed:

* `markers`, `rectangles`, `fills` which are documented on the `plot` function.
* `margin`, `axis` which are barely mentioned in the `Plot` class, but doesn't explain what they do.
* `annotations`, `autoscale` completely undocumented.

Also, some methods will be renamed. For example, the `get_points` method (of the `LineOver1DRangeSeries`, `Parametric2DLineSeries`, ...) will be renamed to `_get_points`: in Python convention, it means that it won't be "public" anymore. Instead, a new public method has been introduced: `get_data`.

**Comment by oscarbenjamin** [comment]
When I try this the attributes are either not there or just None:
```python
In [11]: p = plot(x**2)

In [12]: print(p.markers)
None

In [13]: print(p.rectangles)
None

In [14]: print(p.fills)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-14-84993ed9fccb> in <module>
----> 1 print(p.fills)

AttributeError: 'Plot' object has no attribute 'fills'
```
Do they ever do anything useful?

Why do they need to be removed?

**Comment by Davide-sd** [comment]
Those attributes are used to store numerical data that will be passed to the respective matplotlib functions. Take a look [here](https://github.com/sympy/sympy/blob/master/sympy/plotting/plot.py#L1510)

They can be removed because:
1. they add nothing useful: in fact, they are redundant (see point 2). The backend should facilitate the plotting of symbolic expression. Period.
2. In the new module, once the symbolic part is done the user can retrieve the figure object (a matplotlib figure, or a Plotly figure, ...) and use it to add custom numerical data. No need to add unnecessary wrappers.

**Comment by oscarbenjamin** [comment]
Is it likely that anyone is using these attributes for anything useful?

Is it awkward to continue to provide the attributes even if they are not useful?

(Please don't presume that I understand the plotting code. I barely use the plotting module - I'm asking these questions to help advise how to handle potential BC problems or deprecations.)

**Comment by Davide-sd** [comment]
> Is it awkward to continue to provide the attributes even if they are not useful?

Not only it is awkward, it would likely break the backends, as they require a one-to-one correspondence between the symbolic expression being plotted and the numerical data contained in the figure object. This is especially true for interactive plotting with widgets.

**Comment by oscarbenjamin** [comment]
As I said before please don't presume that I understand anything about the plotting code. This discussion could a lot quicker if you could take the time to explain clearly what you propose to change and why at least for a single example.

Reading between the lines of what you are saying I guess the situation for say `markers` is something like this:

1. It is a feature that currently works to do something useful when using the plotting module with the matplotlib backend.
2. In some way the feature cannot be made to work for other backends that you would like to add.
3. It is not a well designed feature.
4. It is probably not widely used.
5. However in principle something somewhere might depend on it.

The questions then are:

1. Would the new API have an alternative approach for the user to achieve the same end result?
2. How hard would it be to keep the markers attribute there at least just for the matplotlib backend and have it continue to work just as well as it ever did but with a warning suggesting that code be updated to use an alternative approach?

**Comment by Davide-sd** [comment]
> This discussion could a lot quicker if you could take the time to explain clearly what you propose to change and why at least for a single example.

> Would the new API have an alternative approach for the user to achieve the same end result?

With Sympy:

```
from sympy.plotting import plot as plot
# plot a symbolic expression + numerical data
plot(sin(x), markers=[{'args': [[-5, 5], [-1, 1]], 'color': 'r', 'ls': '--'}])
# wonderful the fact that I had to look at the source code to understand how to use it :| Way to go.
```

with the new module you have access to the figure object:

```
from spb import plot
p = plot(sin(x))
p.fig.axes[0].plot([-5, 5], [-1, 1], 'r', ls='--')
# having all the benefits of a quick access to matplotlib docs: help(p.fig.axes[0].plot)
```

Same thing for `annotations`, `fills`, `rectangles`.

> How hard would it be to keep the markers attribute there at least just for the matplotlib backend and have it continue to work just as well as it ever did but with a warning suggesting that code be updated to use an alternative approach?

Surely harder then detecting the use of such keywords and alert the user of an alternative approach.

If I were to add and document those keyword arguments, then users expect them to work on all backends. That's impossible as all plotting libraries are different: one might not have a specific functionality.

Moreover, (in my opinion) the approach of of the current plotting module is no good at all. Why not adding also `bar`, `axline`, `avline`, etc? You would end up mirroring exactly what matplotlib does. Instead, I provide a way of retrieving the figure object, so users can do whatever they want. No unnecessary code in the plotting module: the new plotting module only serves the purpose of quickly visualizing symbolic expression.

So, am I going to invest time and insert those keyword arguments in the new module? No!

**Comment by oscarbenjamin** [comment]
Above you talked about removing "attributes" such as "markers" from the `Plot` class. Your last comment suggests that you are actually talking about removing a keyword argument `markers` from the `plot` function. Or is it both or even more than that?

> > How hard would it be to keep the markers attribute there at least just for the matplotlib backend and have it continue to work just as well as it ever did but with a warning suggesting that code be updated to use an alternative approach?
> 
> Surely harder then detecting the use of such keywords and alert the user of an alternative approach.

I think my confusion here comes from not understanding whether you are referring to attributes or keyword arguments. Also it's not clear what you mean by "alert" here. Do you mean that the call to plot would fail with an exception or that it would still work but with a warning?

> Moreover, (in my opinion) the approach of of the current plotting module is no good at all.

I agree and your suggestion to provide access to the figure object seems like a better way to handle things in general. We're talking about maintaining compatibility (or not) for existing users of the current API though. For that the questions to consider are whether people are likely to be using the feature and if so what would happen to their code if not changed and what changes they could make to their code so that it works as before (or better than before).

As I've said before I don't know the plotting module that well and I don't use it beyond very simple examples so it's hard for me to judge the answers to those questions which is why I'm asking you.

> If I were to add and document those keyword arguments, then users expect them to work on all backends.

Unless I misunderstand the proposal you would not be adding them: they already exist in SymPy. If they are currently undocumented and we don't want to encourage their use then there is no need to document them now. Documentation can be a good way to point out that we don't want people to use something though. For example the documentation could say that `markers` is deprecated and doesn't work with all backends and could also show what is a recommended way to add markers. In fact the `markers` argument to `plot` is currently documented although the docs are not exactly clear and no examples are given:
https://docs.sympy.org/latest/modules/plotting.html#sympy.plotting.plot.plot

**Comment by Davide-sd** [comment]
> Above you talked about removing "attributes" such as "markers" from the Plot class. Your last comment suggests that you are actually talking about removing a keyword argument markers from the plot function. Or is it both or even more than that?

They are the same: in the current sympy plotting module, whatever is specified with the keyword arguments `markers`, `annotations`, etc., gets copied on respective attributes in the `Plot` class. 

> Also it's not clear what you mean by "alert" here. Do you mean that the call to plot would fail with an exception or that it would still work but with a warning?

I would raise a warning telling people the old behavior is not supported, providing an alternative usage example.

> Unless I misunderstand the proposal you would not be adding them: they already exist in SymPy.

They currently exists on sympy, but they do not exist on my module!

**Comment by oscarbenjamin** [comment]
> > Unless I misunderstand the proposal you would not be adding them: they already exist in SymPy.
> 
> They currently exists on sympy, but they do not exist on my module!

Maybe not but here we are talking about merging what you have created into SymPy whose plotting module already exists and is used by many people downstream.

I think I understand what is happening here: The arguments like markers are trivial to support for the matplotlib backend because they just pass parameters directly into matplotlib's APIs but it would be extremely hard to emulate precisely what matplotlib does for any of the other backends. I can see that it would be cleaner not to have these arguments and they shouldn't have been added if the intention was to have multiple backends which is why you niow want to remove them.

However simply removing these is clearly not backwards compatible and it's actually not hard to preserve compatibility here. It shouldn't be difficult just to say that these arguments only work with matplotlib and are deprecated and then list better ways of doing things in the docs.

**Comment by Davide-sd** [comment]
> However simply removing these is clearly not backwards compatible and it's actually not hard to preserve compatibility here. It shouldn't be difficult just to say that these arguments only work with matplotlib and are deprecated and then list better ways of doing things in the docs.

You don't understand the differences between the current plotting module and my module!

The only purpose of the current sympy plotting module is to **draw things only once**, so it doesn't matter where data is stored (symbolic expressions are stored into a `*Series` class; numerical data is stored inside the `Plot` class with `marker`, etc.

My module must also work with interactive widgets, meaning it must support multiple updates! Every object that requires to be plotted must be stored inside a `*Series` class, so the backend class knows exactly how to treat things and in what order. Turns out that Matplotlib is not designed to be interactive! Many Matplotlib objects are not designed to be updated, they must be cleared and replotted. All this to say that, yes, it is very likely to be hard to make that poor design works correctly!

If loosing a few badly documented keyword arguments is something out of you world, than sympy is gonna get two plotting modules.

**Comment by asmeurer** [comment]
Is it possible to enable these old attributes for the use cases of the existing module (with deprecation), but making them not work at all for the instances where they cannot work, like non-matplotlib or interactive?

Another suggestion would be to include both modules in the code as private submodules, like sympy/plotting/_oldplotting/ and sympy/plotting/_newplotting/. By default, the plotting functions would pass through to the _newplotting module, but whenever someone passes a keyword that doesn't work with the new module, it emits a deprecation warning and calls _oldplotting instead. Once the deprecation cycle ends, we would remove _oldplotting and rename _newplotting to just plotting. 

Also, do I understand correctly that attributes don't really have a replacement? In that case, can we deprecate them now, even before we merge in the new module? It would be unfortunate it we can't actually merge in the new module before the deprecation is removed, but if that's the only way, then we might want to shorten the 1 year policy for this case.

**Comment by oscarbenjamin** [comment]
I just took a look at what it would take to implement markers in the `sympy-plot-backends`module and to make the markers parameter to `plot` work you can do something like this:
```diff
diff --git a/spb/backends/matplotlib.py b/spb/backends/matplotlib.py
index ee3fba5..d174d48 100644
--- a/spb/backends/matplotlib.py
+++ b/spb/backends/matplotlib.py
@@ -636,6 +636,16 @@ class MatplotlibBackend(Plot):
                 c = self.ax.fill(x, y, **kw)
                 self._add_handle(i, c, kw)
 
+            elif s.is_deprecated_matplotlib:
+                for marker in s.objects.get('markers', ()):
+                    # make a copy of the marker dictionary
+                    # so that it doesn't get altered
+                    m = marker.copy()
+                    args = m.pop('args')
+                    handles = self.ax.plot(*args, **m)
+                    for h in handles:
+                        self._add_handle(i, h, m)
+
             else:
                 raise NotImplementedError(
                     "{} is not supported by {}\n".format(type(s), type(self).__name__)
diff --git a/spb/functions.py b/spb/functions.py
index 4acb418..be11ceb 100644
--- a/spb/functions.py
+++ b/spb/functions.py
@@ -22,7 +22,7 @@ from spb.series import (
     SurfaceOver2DRangeSeries, ContourSeries, ParametricSurfaceSeries,
     ImplicitSeries, _set_discretization_points,
     List2DSeries, GeometrySeries, Implicit3DSeries,
-    InteractiveSeries
+    InteractiveSeries, DeprecatedMatplotlibSeries
 )
 from spb.utils import (
     _plot_sympify, _check_arguments, _unpack_args, _instantiate_backend
@@ -210,6 +210,10 @@ def _build_line_series(*args, **kwargs):
         else:
             arg = _process_summations(sum_bound, *arg)
             series.append(LineOver1DRangeSeries(*arg[:-1], **kw))
+
+    if kwargs.keys() & DeprecatedMatplotlibSeries.keys:
+        series.append(DeprecatedMatplotlibSeries(**kwargs))
+
     return series
 
 
diff --git a/spb/series.py b/spb/series.py
index b00ec87..98428f6 100644
--- a/spb/series.py
+++ b/spb/series.py
@@ -291,6 +291,10 @@ class BaseSeries:
     # If True, the backend will attempt to render it on a polar-projection
     # axis, or using a polar discretization if a 3D plot is requested
 
+    is_deprecated_matplotlib = False
+    # Represents deprecated features that only work with the matplotlib
+    # backend such as markers.
+
     use_cm = True
     # Some series might use a colormap as default coloring. Setting this
     # attribute to False will inform the backends to use solid color.
@@ -3088,3 +3092,15 @@ class GeometryInteractiveSeries(GeometrySeries, InteractiveSeries):
     def __str__(self):
         s = super().__str__()
         return "interactive " + s + " with parameters " + str(tuple(self._params.keys()))
+
+
+class DeprecatedMatplotlibSeries(BaseSeries):
+    """Holds instructions for doing deprecated things that only work with the
+    matplotlib backend"""
+
+    keys = {'markers'}
+
+    is_deprecated_matplotlib = True
+
+    def __init__(self, **kwargs):
+        self.objects = {k: kwargs[k] for k in self.keys if k in kwargs}
```
With that you can do:
```python
In [1]: from spb import plot

In [2]: plot(x**2, markers=[{'args': [[-5, 5], [-1, 1]], 'color': 'r', 'ls': '--'}])
Out[2]: <spb.backends.matplotlib.MatplotlibBackend at 0x7f5bfc44bb20>
```
The same `DeprecatedMatplotlibSeries` class could be used to handle the other parameters like `rectangles` and `fills` and possibly the others as well. It could also be changed to give a warning for each of the used parameters. Maybe there is some situation e.g. in an animation where that wouldn't work and the logic around `_add_handle` needs to be a bit cleverer. I think it should be possible to at least make the basic cases work though and it's also possible to give a warning.

**Comment by Davide-sd** [comment]
> Another suggestion would be to include both modules in the code as private submodules, like sympy/plotting/_oldplotting/ and sympy/plotting/_newplotting/. By default, the plotting functions would pass through to the _newplotting module, but whenever someone passes a keyword that doesn't work with the new module, it emits a deprecation warning and calls _oldplotting instead. Once the deprecation cycle ends, we would remove _oldplotting and rename _newplotting to just plotting.

I'm afraid that's not a good solution, because the new module introduces new keyword arguments that are not used on the old one. So, if a user writes a command combining new and old keyword arguments, you would likely end up with unsatisfactory results.

> Also, do I understand correctly that attributes don't really have a replacement? In that case, can we deprecate them now, even before we merge in the new module? It would be unfortunate it we can't actually merge in the new module before the deprecation is removed, but if that's the only way, then we might want to shorten the 1 year policy for this case.

Before discussing this, I think it's better to fully understand what is going to change with the new module. I took the time to compare the two modules and these are the finding:

Keyword arguments that are used in `sympy.plotting` but I removed on the new module:

* `axis`: this is undocumented. It removes the axis from the figure. Difficult if not impossible to implement on other backends.
* `autoscale`: no idea what this does, as it was undocumented. Maybe it was used to deal with limitations of Matplotlib back in the days when the module was developed.
* `margin`: undocumented. Difficult if not impossible to implement on other backends.
* `annotations`, `markers`, `rectangles`, `fill`: discussed above.

Keyword arguments that have been renamed to improve typing experience. If necessary, it is possible to intercept the old one and raise a deprecation message.

* `aspect_ratio` becomes `aspect`
* `nb_number_of_points` becomes `n`
* `nb_number_of_points_x` becomes `n1`
* `nb_number_of_points_y` becomes `n2`

Classes and functions that have been removed:

* `centers_of_segments`, `centers_of_faces`, `flat`: they are used by the old adaptive algorithm. They are no longer necessary. Those functions can be copy-paste and insert a deprecation message.
* `_matplotlib_list`: this has been moved into `MatplotlibBackend` because only this backend uses such functionality.
* `BaseBackend`: no longer needed. There's nothing that comes close to such concept.
* `plot_backends`: a dictionary mapping string to backend classes. No longer needed on the new module.
* `DefaultBackend`: a class that returns the default backend. No longer needed on the new module, as the implementation is massively different.
* `TextBackend`: create text-based plots on terminals. Removed because (in my humble opinion) it is unable to provide a correct representation of a function.

Methods from the `*Series` classes that have been renamed. It should be straightforward to recreate these methods and add a deprecation warning:

* The `get_points` method from `LineOver1DRangeSeries`, `List2DSeries`, `Parametric2DLineSeries`, `Parametric3DLineSeries`: renamed to `_get_points`. Both methods returns points that need to be further post-processed. The change in name is meant to highlight the fact that it shouldn't be used publicly.
* The `get_parameter_points` method from `Parametric2DLineSeries`, `Parametric3DLineSeries` and the `get_parameter_meshes` from the `ParametricSurfaceSeries`: renamed to `_discretize`. This change allows to better reuse code from the parent classes.
* The `get_meshes` method from `SurfaceOver2DRangeSeries`, `ContourSeries`, `ParametricSurfaceSeries`: renamed to `get_data`, which is the only public method on the new module to obtain numerical data.
* The `get_raster` method from `ImplicitSeries`: renamed to `get_data`, which is the only public method on the new module to obtain numerical data.

Methods from the `*Series` classes that have been removed as they are no longer needed:

* `get_segments`  from `Line2DBaseSeries`. It has been moved into `MatplotlibBackend` because only matplotlib needs such functionality.
* `get_color_array` from the `Line2DBaseSeries`, `SurfaceBaseSeries`.

**Comment by asmeurer** [comment]
> I'm afraid that's not a good solution, because the new module introduces new keyword arguments that are not used on the old one. So, if a user writes a command combining new and old keyword arguments, you would likely end up with unsatisfactory results.

Why is that a problem? The only point is to continue to support existing code that uses the old keyword arguments (with a depreciation warning). Existing code won't use them and new code shouldn't use them since they are deprecated, so combining the two can easily be an error.

**Comment by asmeurer** [comment]
> aspect_ratio becomes aspect

I guess matplotlib does this too, but I'm not personally a fan of it. I know what "aspect ratio" means, but to me, "aspect" by itself does not mean aspect ratio. If you lookup "aspect" in the dictionary none of the definitions are the same as "aspect ratio", because they aren't synonyms. Personally, when I see just "aspect" in matplotlib's API, I don't immediately know what it means because of this (it could, in principle, mean many different things), whereas "aspect ratio" is unambiguous and clear.

**Comment by asmeurer** [comment]
> get_segments from Line2DBaseSeries. It has been moved into MatplotlibBackend because only matplotlib needs such functionality.

Just to be clear, get_segments has already been deprecated for some time (https://docs.sympy.org/latest/explanation/active-deprecations.html#the-get-segments-attribute-of-plotting-objects), so there's no issue with removing it.

**Comment by James-G-Hill** [comment]
Any progress on this? Having used the new spb module as a separate import I can see that it is a substantial advance on the existing plotting module. What would need to happen to get this issue moving?

**Comment by Davide-sd** [comment]
Hello @James-G-Hill ,

Lately I've been working on improving the documentation, which required to implement new Sphinx extensions in order to create screenshots of interactive plots. While the html version is at a very good point (you can explore the [latest build here](https://sympy-plot-backends.readthedocs.io/en/latest/overview.html#examples)), I still have to attempt to build a PDF version.

Speaking of merging: I'm not really sure if it is a good idea to include this module into SymPy because the documentation is massive:

* currently, a single build requires about 15 minutes on a ReadTheDocs server.
* and at least 400MB of disk space for the html version.

That's a lot... :)

**Comment by oscarbenjamin** [comment]
There are basically two approaches to merging something like this:

1. Apply changes in pieces.
2. Copy the new module in wholesale.

I'm guessing that 1. has become too difficult at this stage because the codebases have diverged for so long but that might not be true. To follow approach 1 what we need is one or more PRs that change the existing plotting module so that the structure there matches up with the basic structure of `spb`. Then the different `spb` backends or features can just be added either one by one or all together.

Following 2. means that `spb` (or a fork of `spb`) first needs to be made more compatible with the existing plotting module. @Davide-sd made a list of known incompatibilities above. Some of those changes are probably fine. Others could easily be made compatible but it would be better to do that first outside the sympy repo before merging in all at once. I showed a diff above that demonstrates how that could be done e.g. for the `markers` parameter. That could be extended to handle most of the keyword arguments that were changed or removed but that would need to happen in `spb` first.

**Comment by oscarbenjamin** [comment]
> Speaking of merging: I'm not really sure if it is a good idea to include this module into SymPy because the documentation is massive:
> 
> * currently, a single build requires about 15 minutes on a ReadTheDocs server.
> * and at least 400MB of disk space for the html version.

There are ways to work around that e.g. the docs could be in a separate repo.

**Comment by James-G-Hill** [comment]
Thanks for your replies, I don't have much skill to help I'm afraid, I have only dabbled in many areas of sympy. But I have encountered many problems with the current plotting module so am aware of how much the inclusion of this new module would improve the user experience. The current module for plotting has a lot of issues that @Davide-sd has overcome and he's also added many new features which make it work in an intuitive way. So I really wanted to raise this because I think @Davide-sd 's work is excellent.

**Comment by oscarbenjamin** [comment]
> So I really wanted to raise this because I think @Davide-sd 's work is excellent.

I don't think there's any disagreement about that. The missing work is the part that integrates this into SymPy (if that is the intended goal).

**Comment by asmeurer** [comment]
We will need to figure out something for the documentation. Currently our docs are hosted on GitHub pages and 400MB per build will cause the docs repo to quickly fill up. Furthermore, it could break things like the CircleCI docs build. 

We'll need to come up with something with the images can be hosted somewhere else. If we have to pay for hosting it is a possibility, as long as it isn't too much. 

I would see if there are other projects that have similar issues (docs with a lot of images) and see how they solve them. If we can figure out the other issues here, I don't want the docs problem to block the merge, so if we can't find something easy, we can just disable the images in the docs for now. 

> which required to implement new Sphinx extensions in order to create screenshots of interactive plots.

Can you not include the interactive plots directly? Or do they require a server to run?

**Comment by Davide-sd** [comment]
> Can you not include the interactive plots directly? Or do they require a server to run?

Interactive plots do not requires a server running. They are just html files containing the data which will be rendered by some JS library. They are only meant to better show the capabilities of a plotting library.

I have just implemented options to the Sphinx extensions to exclude the html files from the build. Hence, only static pictures will be included in the docs. By doing so the final build is around 150MB.

**Comment by michelececcacci** [comment]
Hi Davide, i want to work on improving the plotting module,  and i have been told to talk to you by @asmeurer. Are there any tasks i can do to get started? I started by looking at ideas in the GSOC markdown document. My original plan was to help porting the pyglet module, but i got told it's  not that important, and that there are better plotting backends we could be using. Can you give me a few pointers for that? Thanks in advance!

**Comment by Davide-sd** [comment]
Hello @michelececcacci , I'll write an email to you.... :)

**Comment by TiloRC** [comment]
Hello. I'm considering writing my GSOC proposal on something to do with plotting—perhaps I could help merge this new plotting module into SymPy? I see that [michelececcacci](https://github.com/michelececcacci) is also interested in this. Would it be possible for both of us to work on plotting related projects? If so, it would be nice to know what michelececcacci is planning to work on so that I could write a proposal that didn't overlap with theirs.

**Comment by vikasprajapati-01** [comment]
Hi @Davide-sd and @oscarbenjamin ,
I am planning to apply for the "Improve the plotting module" GSoC 2026 project. I have reviewed the history of this issue and the major PRs from 2023 — specifically #25469 (Series refactor + lambdify migration), #25462 (backend refactoring), #25414 (unified get_data), and #25120 (deprecation of nb_of_points_*).

Since the core structural merge is largely complete, I would like to confirm: is the current priority for a 2026 GSoC project focused on:

Interactive module fixes — solving the Panel/Param "parameter collision" bug and integrating plotgrid with interactive apps?
New features — Phase portraits for ODEs, 2D/3D linear operators, animations, and improved plot_parametric_region?
Auto region-of-interest detection for smarter plot limits?
Conda packaging — debugging and fixing the k3d/vtk dependency build failures?

I have the spb repository cloned locally and have already explored series_2d_3d.py and the backend abstraction layer. I'd love your guidance on which items you consider highest priority.
